servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert subscription != null;
+
+                if (src == null) {
+                    return;
+                }
+
+                // onNext will produce AT-MOST N items (as received)
+                if (!channel.isOpen()) {
+                    throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
+                }
+
+                channel.writeInbound(toByteBuf(src));
+                Buffer buffer = drainChannelInToSingleBuffer(channel, allocator);
+
+                if (buffer != null && buffer.readableBytes() > 0) {
+                    subscriber.onNext(buffer);
+                } else {
+                    // Not enough data to decompress, ask for more
+                    subscription.request(1);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                safeCleanup(channel);
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                cleanup(channel);
+                subscriber.onComplete();
+            }
+        });
+    }
+
+    private EmbeddedChannel newEmbeddedChannel(final ChannelHandler handler, final BufferAllocator allocator) {
+        final EmbeddedChannel channel = new EmbeddedChannel(handler);
+        channel.config().setAllocator(getByteBufAllocator(allocator));
+        return channel;
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer) {
+        return BufferUtils.toByteBuf(buffer);
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer, final int offset, final int length) {
+        return BufferUtils.toByteBuf(buffer.readerIndex(buffer.readerIndex() + offset).readSlice(length));
+    }
+
+    private static void preparePendingData(final EmbeddedChannel channel) {
+        try {
+            channel.close().syncUninterruptibly().get();
+            channel.checkException();
+        } catch (InterruptedException | ExecutionException ex) {
+            LOGGER.error("Unexpected exception from EmbeddedChannel", ex);
+        }
+    }
+
+    @Nullable
+    private static Buffer drainChannelInToSingleBuffer(
+            final EmbeddedChannel channel, final BufferAllocator allocator) {
+        if (channel.inboundMessages().isEmpty()) {
+            return null;
+        }
+
+        if (channel.inboundMessages().size() == 1) {
+            return newBufferFrom(channel.readInbound());
+        } else {
+            CompositeBuffer compositeBuffer = allocator.newCompositeBuffer();
+            ByteBuf part;
+            while ((part = channel.readInbound()) != null) {
+                compositeBuffer.addBuffer(newBufferFrom(part));
+            }
+
+            return compositeBuffer;
+        }
+    }
+
+    @Nullable
+    private static Buffer drainChannelOutToSingleBuffer(
```

Comment:
Can we share code between `drainChannelInToSingleBuffer` and `drainChannelOutToSingleBuffer`? A common method can just take a `Queue<Object>`